File: BensBible2/Models/ExerciseState.swift
import Foundation

enum ExerciseType: String {
    case fillBlank
    case wordDrag
    case multipleChoice
}

enum FillBlankSegment {
    case word(String)
    case blank(answer: String)
}

enum ExerciseState {
    case fillBlank(verse: MemorizedVerse, segments: [FillBlankSegment])
    case wordDrag(verse: MemorizedVerse, shuffledWords: [String], correctOrder: [String])
    case multipleChoice(verse: MemorizedVerse, options: [String])

    var verse: MemorizedVerse {
        switch self {
        case .fillBlank(let v, _): return v
        case .wordDrag(let v, _, _): return v
        case .multipleChoice(let v, _): return v
        }
    }

    var exerciseType: ExerciseType {
        switch self {
        case .fillBlank: return .fillBlank
        case .wordDrag: return .wordDrag
        case .multipleChoice: return .multipleChoice
        }
    }
}

File: android/app/src/main/java/com/bensbible/app/model/ExerciseState.kt
package com.bensbible.app.model

import com.bensbible.app.data.MemorizedVerseEntity

sealed class ExerciseState {
    abstract val verse: MemorizedVerseEntity

    data class FillBlank(
        override val verse: MemorizedVerseEntity,
        val segments: List<FillBlankSegment>
    ) : ExerciseState()

    data class WordDrag(
        override val verse: MemorizedVerseEntity,
        val shuffledWords: List<String>,
        val correctOrder: List<String>
    ) : ExerciseState()

    data class MultipleChoice(
        override val verse: MemorizedVerseEntity,
        val options: List<String>
    ) : ExerciseState()
}

sealed class FillBlankSegment {
    data class Word(val text: String) : FillBlankSegment()
    data class Blank(val answer: String) : FillBlankSegment()
}

enum class ExerciseType {
    FILL_BLANK, WORD_DRAG, MULTIPLE_CHOICE
}
